[OpenACC][CIR] Add parallelism determ. to all acc.loops

erichkeane opened the pull request:
PR #143720 adds a requirement to the ACC dialect that every acc.loop must have a seq, independent, or auto attribute for the 'default' device_type. The standard has rules for how this can be intuited:

orphan/parallel/parallel loop: independent
kernels/kernels loop: auto
serial/serial loop: seq, unless there is a gang/worker/vector, at which point it should be 'auto'.

This patch implements all of this rule as a 'cleanup' step on the IR generation for combined/loop operations. Note that the test impact is much less since I inadvertently have my 'operation' terminating curley matching the end curley from 'attribute' instead of the front of the line, so I've added sufficient tests to ensure I captured the above.

erichkeane committed:
[OpenACC][CIR] Add parallelism determ. to all acc.loops

PR #143720 adds a requirement to the ACC dialect that every acc.loop
must have a seq, independent, or auto attribute for the 'default'
device_type. The standard has rules for how this can be intuited:

orphan/parallel/parallel loop: independent
kernels/kernels loop: auto
serial/serial loop: seq, unless there is a gang/worker/vector, at which
point it should be 'auto'.

This patch implements all of this rule as a 'cleanup' step on the IR
generation for combined/loop operations. Note that the test impact is
much less since I inadvertently have my 'operation' terminating curley
matching the end curley from 'attribute' instead of the front of the
line, so I've added sufficient tests to ensure I captured the above.

razvanlupusoru commented on clang/lib/CIR/CodeGen/CIRGenStmtOpenACCLoop.cpp:
```diff
@@ -22,6 +22,63 @@ using namespace clang::CIRGen;
+  assert(dk == OpenACCDirectiveKind::Serial ||
+         dk == OpenACCDirectiveKind::SerialLoop);
+
+  if (op.getWorkerAttr() || op.getVectorAttr() || op.getGangAttr()) {
```
This is an acc dialect problem - but basically we encode in a different fields whether a loop has `gang` or `gang(value)` or `gang(dim:)`. The more complete check is:
```
    bool hasDefaultGangWorkerOrVector =
        loopOp.hasVector() || loopOp.getVectorValue() || loopOp.hasWorker() ||
        loopOp.getWorkerValue() || loopOp.hasGang() ||
        loopOp.getGangValue(mlir::acc::GangArgType::Num) ||
        loopOp.getGangValue(mlir::acc::GangArgType::Dim) ||
        loopOp.getGangValue(mlir::acc::GangArgType::Static);
```
And probably it should be in a utility in acc dialect itself.

erichkeane committed:
Correct /simplify

The gang/worker/vector check was insufficient based on review, so this
fixes it.  It also moves the check for the ParallelismFlag and
gang/worker/vector check to LoopOp so that it can be used elsewhere. We
also can simplify the Clang version here for the same reasons.

razvanlupusoru reviewed (approved):
Thank you!
